### PR TITLE
pipeline memory efficiency using pool

### PR DESCRIPTION
**Pull request by geyslan**
### 1. Explain what the PR does

04bf8534f **perf:(pipeline) memory efficiency using pool** _<sub>(2023/jul/01) Geyslan Gregório \<geyslan@gmail.com\></sub>_

```
This commit employs sync.Pool to bolster memory performance in event
handling. The benchmarking file, events_pipeline_bench_test.go,
simulates the pipeline execution and measures sync.Pool's effectiveness.

Benchmark command:

go test \
  -benchmem -benchtime=100x \
  -run=^$ -bench ^(BenchmarkGetEventFromPool|BenchmarkNewEventObject)$ \
  github.com/aquasecurity/tracee/pkg/ebpf

Findings:

Environment:
- OS: Linux
- Arch: amd64
- Package: github.com/aquasecurity/tracee/pkg/ebpf
- CPU: Intel Core i7-12700H

Benchmarking Results:

BenchmarkGetEventFromPool (with pooling):
- Time per Op: 1740.88 ms
- Memory per Op: 243.5 KB
- Allocations per Op: 543

BenchmarkNewEventObject (without pooling):
- Time per Op: 1754.69 ms
- Memory per Op: 896.2 KB
- Allocations per Op: 2001

Using object pooling reduces memory allocation and allocation count
by ~3.7x.

In conclusion, sync.Pool results in remarkable memory optimization
without substantial impact on runtime. This efficiency is critical in
the pipeline that must have high throughput, as it can facilitate
garbage collection and promote resource efficiency.

For more information, see the benchmarking results of tracee actually
running on #3297.
```

### CODE REVIEW

**Comment by geyslan** [comment]
Since it's difficult to call `decodeEvents()` directly without simulating all prior steps, the benchmark tests were implemented in a separate file, `events_pipeline_bench_test.go`, which simulates the pipeline execution and measures the performance of the use of `sync.Pool`.

Results:

```
go test \
  -benchmem -benchtime=10000x -cpu=1 \
  -run=^$ -bench ^(BenchmarkEventPool|BenchmarkEventNew)$ \
  github.com/aquasecurity/tracee/pkg/ebpf
```

| Test # | EventPool ns/op | EventNew ns/op | % Improvement in Time | EventPool B/op | EventNew B/op | % Improvement in Memory | EventPool allocs/op | EventNew allocs/op |
|--------|-----------------|----------------|----------------------|----------------|---------------|------------------------|---------------------|--------------------|
| 1      | 1,178,529       | 1,184,807      | +0.53%               | 26             | 448           | +94.20%                | 0                   | 1                  |
| 2      | 1,187,255       | 1,175,401      | -1.01%               | 26             | 448           | +94.20%                | 0                   | 1                  |
| 3      | 1,186,611       | 1,169,843      | -1.43%               | 26             | 448           | +94.20%                | 0                   | 1                  |
| 4      | 1,184,051       | 1,196,323      | +1.03%               | 26             | 448           | +94.20%                | 0                   | 1                  |
| 5      | 1,180,822       | 1,182,371      | +0.13%               | 26             | 448           | +94.20%                | 0                   | 1                  |
| 6      | 1,181,872       | 1,186,390      | +0.38%               | 26             | 448           | +94.20%                | 0                   | 1                  |

**Comment by geyslan** [comment]
Accumulating results of 200 * N (`-benchtime=100x`), which keeps the simulated pipeline warm, e.g.:

```go
	go func() {
		for i := 0; i < b.N; i++ {
			for j := 0; j < 200; j++ {
				ctxChan <- &ctx
			}
		}
	}()
```

Benchmark Name    | Time per Operation (ns/op) | Total Elapsed Time (seconds) | Memory Allocation (B/op) | Memory Allocation Improvement (%) | Number of Allocations (allocs/op) | Allocation Improvement (%)
----------------- | -------------------------- | ----------------------------- | ------------------------ | --------------------------------- | ------------------------------- | --------------------------
BenchmarkEventNew | 237,529,472                | ~23.75                        | 89,607                   | Reference                         | 200                             | Reference
BenchmarkEventPool| 236,495,431                | ~23.65                        | 2,648                    | ~97.05% improvement               | 0                               | 100% improvement

**Comment by NDStrahilevitz** [comment]
Nice work! I toyed with this optimization before but was not able to benchmark a difference, nice to see it has an effect. If and when memory arenas (added in go 1.20) are moved to standard from experimental, we may be able to further optimize event allocations.

File: pkg/ebpf/events_pipeline.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ func (t *Tracee) handleEvents(ctx context.Context) {
 	var errcList []<-chan error
 
 	// Source pipeline stage.
-	eventsChan, errc := t.decodeEvents(ctx, t.eventsChannel)
+	eventsChan, evtPool, errc := t.decodeEvents(ctx, t.eventsChannel)
```
Comment (NDStrahilevitz) [inline pkg/ebpf/events_pipeline.go:34]: WDYT about moving the pool to the tracee struct so it won't need to be passed per pipeline stage? Later when #3254 is tackled, the pool should belong to the pipeline.

**Review by NDStrahilevitz — COMMENTED** [review]
I see that the pool isn't passed to some of the stages (like sorting, enrichment, etc.), I assume because events aren't dropped there right? Have you tested with these additional pipeline stages (at least enrichment since we enable it by default in kubernetes).

**Comment by geyslan** [comment]
> I see that the pool isn't passed to some of the stages (like sorting, enrichment, etc.), I assume because events aren't dropped there right?

Exactly.

> Have you tested with these additional pipeline stages (at least enrichment since we enable it by default in kubernetes).

So far, I've only simulated (roughly) the pipeline with the events_pipeline_bench_test.go. As you asked, I'll generate pyroscope outputs to take a look in the wild.

File: pkg/ebpf/events_pipeline.go (hunk 1 above)
Comment (geyslan) [reply pkg/ebpf/events_pipeline.go:34]: Good idea. I've just did that. Pushing changes soon.

**Comment by geyslan** [comment]
## Test

```shell
#!/bin/bash

sudo ./dist/tracee --metrics --pprof --pyroscope -f e=sched_process_exec -f comm=who -o none &

sleep 90

i=0; while ((i < 700000)); do who; ((i++)) ; done

sleep 60

sudo pkill tracee
```

## main branch

image



## event-pool branch:

image


## Summary

- Less mallocs 1-(39133857/39752539) ~= 1,55%
- Heap (in use and objects) seem to be used less in average.
- GC pressure data (mean) appears to be a bit higher, but we have to consider the initial peak from the event pool warming.
- GC pressure time (mean) shows an improvement of 1-(6,84/7,44)µs ~= 8,06%

In the event-pool branch heap graphs we can notice more sequential peaks and valleys.

**Comment by geyslan** [comment]
The previous test is tainted. The warming of the event pool is useless since we are waiting 90s after start tracee to initiate the event emission. The pool cools down once gc is triggered in meanwhile.

So, let's test without warming the pool.

## Test

```shell
#!/bin/bash

sleep 10 # cool down after building

sudo ./dist/tracee --metrics --pprof --pyroscope -f e=sched_process_exec -f comm=who -o none &

sleep 5

i=0; while ((i < 700000)); do who; ((i++)) ; done

sleep 30

sudo pkill tracee
```

## main branch

image

## event-pool branch

image

## Results

- Less mallocs 1-(38707896/39296721) ~= `-1,49%`
- Heap (in use and objects) seem to be used more in average:
  - Less malloc rate 1-(50049/55230) ~= `-9,38%`
  - More heap objects in use 1-(206973/182242) ~= `+13,57%` (in this case is good)
- GC pressure data (avg) appears to be a bit lower: `53.3MB`
- GC pressure time (avg) shows an improvement of 1-(8,4/9,47)µs ~= `-11,29%`

We can see in the malloc, free and heap graphs less volatility, which is the purpose of using the event pool.

**Review by NDStrahilevitz — APPROVED** [review]
Needs a cleanup of leftovers. You can do the function move by your style call, otherwise LGTM.

File: pkg/ebpf/events_pipeline.go [hunk 2]
```diff
@@ -218,49 +218,58 @@ func (t *Tracee) decodeEvents(outerCtx context.Context, sourceChan chan []byte)
...
+			evt.ProcessName = string(bytes.TrimRight(ctx.Comm[:], "\x00"))
+			evt.HostName = string(bytes.TrimRight(ctx.UtsName[:], "\x00"))
+			evt.CgroupID = uint(ctx.CgroupID)
+			evt.ContainerID = containerData.ID
+			evt.Container = containerData
+			evt.Kubernetes = kubernetesData
+			evt.EventID = int(ctx.EventID)
+			evt.EventName = eventDefinition.Name
+			evt.MatchedPoliciesKernel = ctx.MatchedPolicies
+			// evt.MatchedPoliciesUser = 0
```
[the remark this answers is not in the document]
Comment (geyslan) [reply pkg/ebpf/events_pipeline.go:249]: I left these assignments commented out to emphasize that they aren't necessary at this step and would only degrade performance. But I will remove them making the above comment clearer.

[the remark this answers is not in the document]
Comment (yanivagman) [reply pkg/ebpf/events_pipeline.go:249]: Since these values are not intially assigned, is there any chance that we will get wrong values from previous events that used this struct?

File: pkg/ebpf/events_pipeline.go [hunk 3]
```diff
@@ -218,49 +218,58 @@ func (t *Tracee) decodeEvents(outerCtx context.Context, sourceChan chan []byte)
...
-				EventName:             eventDefinition.Name,
-				MatchedPoliciesKernel: ctx.MatchedPolicies,
-				ArgsNum:               int(argnum),
-				ReturnValue:           int(ctx.Retval),
-				Args:                  args,
-				StackAddresses:        stackAddresses,
-				ContextFlags:          flags,
-				Syscall:               syscall,
-			}
+			// get an event pointer from the pool
```
[the remark this answers is not in the document]
Comment (geyslan) [reply pkg/ebpf/events_pipeline.go:221]: Do you mean only the `evt := t.eventsPool.Get().(*trace.Event)`? If so, we should wrap the `t.eventsPool.Put(evt)` as well.

If you mean the assignments, I thought about that, but the function would have so many args:
- *bufferdecode.Context
- *tracee.Container
- *tracee.Kubernetes
- *events.Event (name)
- argnum
- []tracee.Argument
- []uint64 (stackAddresses)
- tracee.ContextFlags
- syscall (string).

So I left it as it is.

File: pkg/ebpf/events_pipeline.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (geyslan) [reply pkg/ebpf/events_pipeline.go:249]: > Since these values are not intially assigned, is there any chance that we will get wrong values from previous events that used this struct?

Only if the next stages don't assign them.

[the remark this answers is not in the document]
Comment (NDStrahilevitz) [reply pkg/ebpf/events_pipeline.go:249]: > Only if the next stages don't assign them.

Then we should probably zero these just to be sure. It's premature optimization not to do so IMO. And, if we ever add new code which assumes the fields are correctly zeroed, this would be easy to miss.

File: pkg/ebpf/events_pipeline.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (NDStrahilevitz) [reply pkg/ebpf/events_pipeline.go:221]: Fair enough, then let's leave as is.

**Comment by rafaeldtinoco** [comment]
Can I give a quick look before this is merged ?

**Comment by geyslan** [comment]
> Can I give a quick look before this is merged ?

For sure.

e2e 980 is green.

**Comment by geyslan** [comment]
As discussed offline with @NDStrahilevitz, there are other stages in the pipeline that can also make use of the event pool, such as instantiating derived and subscription events.

File: pkg/ebpf/events_pipeline.go [hunk 4]
```diff
@@ -218,52 +218,57 @@ func (t *Tracee) decodeEvents(outerCtx context.Context, sourceChan chan []byte)
...
+			evt.MatchedPoliciesKernel = ctx.MatchedPolicies
+			evt.MatchedPoliciesUser = 0
+			evt.MatchedPolicies = []string{}
+			evt.ArgsNum = int(argnum)
+			evt.ReturnValue = int(ctx.Retval)
+			evt.Args = args
+			evt.StackAddresses = stackAddresses
+			evt.ContextFlags = flags
+			evt.Syscall = syscall
+			evt.Metadata = nil
```
Comment (NDStrahilevitz) [inline pkg/ebpf/events_pipeline.go:253]: Why `nil` and not `&trace.Metadata{}` as before?

Comment (geyslan) [reply pkg/ebpf/events_pipeline.go:253]: To be discussed, but what I think is that the current instantiation `evt := trace.Event{`, sets Metadata to nil. `&trace.Metadata{}` doesn't initialise its map Properties field `map[string]interface{}`. If we go that way to avoid nil pointer deref and so, that can be also instantiated prematurely and unnecessarily if we instantiate Metadata fully. WDYT?

Comment (NDStrahilevitz) [reply pkg/ebpf/events_pipeline.go:253]: Actually it may be better to leave it this way, since we define it as `omitempty` in the `json` tag of trace.Event.

**Review by rafaeldtinoco — APPROVED** [review]
LGTM. I think this was a great example of PR. The benchmark was theoretical but yet ingenious. Nice work @geyslan!
